### PR TITLE
4.x decimal type

### PR DESCRIPTION
**Pull request by ADmad**
Refs #10233

### CODE REVIEW

**Comment by ADmad** [comment]
Not sure why `DecimalType` supports parsing of localized string. Anyone using localized string for decimals is asking for trouble.

**Comment by dereuromark** [comment]
Well, the input format for it can differ per language. We use a comma not a dot. As such this was always a usability issue to allow for this.

**Comment by ADmad** [comment]
@dereuromark Yeah I figured. Though `Number::parseFloat()` is used for parsing which as the name suggests returns floats so you would end up with same problems too large / small values.

**Comment by ADmad** [comment]
It seems sqlite is unable to deal with large numers https://travis-ci.org/cakephp/cakephp/jobs/515170690#L674. Same test passes for other dbs. Guess I'll just skip the test for sqlite.

**Comment by inoas** [comment]
I don't think the DecimalType should support localization.
A LocalizedDecimalType could but for writing between the database and PHP a string and a dot for separation sounds consistent, expected and reliable.

File: src/Database/Type/DecimalType.php [hunk 1]
```diff
@@ -45,50 +45,55 @@ class DecimalType extends BaseType implements BatchCastingInterface
...
-        if (!is_scalar($value)) {
-            throw new InvalidArgumentException(sprintf(
-                'Cannot convert value of type `%s` to a decimal',
-                getTypeName($value)
-            ));
-        }
-        if (is_string($value) && is_numeric($value)) {
+
+        if (is_numeric($value)) {
             return $value;
```
Comment (markstory) [inline src/Database/Type/DecimalType.php]: Should this string cast float/ints into string?

File: tests/TestCase/Database/Type/DecimalTypeTest.php [hunk 2]
```diff
@@ -177,6 +194,12 @@ public function testMarshal()
 
         $result = $this->type->marshal(['3', '4']);
         $this->assertNull($result);
+
+        // This test is to indicate the problem that will occur if you use
+        // float/double values which get converted to scientific notation by PHP.
+        // To avoid this problem always using strings to indicate decimals values.
+        $result = $this->type->marshal(1234567890123456789.2);
```
Comment (markstory) [inline tests/TestCase/Database/Type/DecimalTypeTest.php]: Marshalling should generally get strings, so this situation should be rare. Should there be a test added for values with 10+ of decimal places not losing precision?

**Comment by markstory** [comment]
@ionas @ADmad removing localized float support will be interpreted as a bug/defect by a few people. It was added after issues were opened about localized values being corrupted.

**Comment by ADmad** [comment]
> I don't think the DecimalType should support localization.

@ionas It's an opt in feature which has to be explicitly enable. So those who don't need it are safe.

File: src/Database/Type/DecimalType.php (hunk 1 above)
Comment (ADmad) [reply src/Database/Type/DecimalType.php]: We already specify the binding type as `PDO::PARAM_STR` through the `toStatement()`. So there's no need for explicit casting, let PDO handle the value conversion as needed.

File: tests/TestCase/Database/Type/DecimalTypeTest.php (hunk 2 above)
Comment (ADmad) [reply tests/TestCase/Database/Type/DecimalTypeTest.php]: I have added more tests.

**Comment by dereuromark** [comment]
I would also be fine if the core shipped with two separate Type classes: Decimal and LocalizedDecimal. The latter could extend the first. We might need sth similar for other types, too?

**Comment by ADmad** [comment]
I don't think a separate class to handle localised values is necessary. Supporting localised values only needs few lines of extra code and doesn't add any significant complexity.

**Comment by ADmad** [comment]
@markstory Anything else to be done here?

**Pull request merged**